zerver/models.py

```diff
@@ -320,11 +332,20 @@ def get_realm_by_email_domain(email):
     if settings.REALMS_HAVE_SUBDOMAINS:
         raise GetRealmByDomainException(
             "Cannot get realm from email domain when settings.REALMS_HAVE_SUBDOMAINS = True")
-    try:
-        alias = RealmAlias.objects.select_related('realm').get(domain = email_to_domain(email))
+    domain = email_to_domain(email)
+    query = RealmAlias.objects.select_related('realm')
+    alias = query.filter(domain=domain, subdomains_allowed=False).first() or \
```

Comment:
If there are two aliases: (realm1, 'example.com', True) and (realm2, 'example.com', False) then a query for 'example.com' should return realm2. What I am doing here is that I first query for an alias having `domain` set to 'example.com' and `subdomains_allowed` to `False`. If the result is not `None` then it is returned otherwise a second query is made for an alias having `domain` set to 'example.com' and `subdomains_allowed` to `True`. If the result is not `None` then it is returned otherwise if it is also `None` then we go on to check if any of the parent domains matches.
P.S.: I have made all other requested changes so you can have a look at them too whether they are fine or not(I have kept the changes in a separate commit for easing the review). :)